### ISSUE TITLE
Clearing cache does not seem to free memory

### ISSUE THREAD

**Issue by reubengann** [pauldendulk/Mapsui#190]
If I have a tile map loaded via MbTilesTileSource on a TileLayer, and I zoom and pan the map, I can see the memory usage going up, as one would expect. I then load a different map, and before doing this I try to clear out the old data, which I've attempted to do in multiple ways (by clearing the layers, calling dispose on the map, or by invoking ClearCache):
```
private void LoadMap(string location)
{
    MyMapControl.Map.Layers.Clear();
    MyMapControl.Map.Dispose();
    MyMapControl.Map.ClearCache();
    _source = new MbTilesTileSource(new SQLiteConnectionString(PathLibrary[location], false));
    MyMapControl.Map.Layers.Add(new Mapsui.Layers.TileLayer(_source));
    MyMapControl.Refresh();
}
```
The thing is, the memory usage never decreases. Loading a new map and panning, zooming, increases the memory, and never decreases it. The profiler indicates that nearly all of the memory is in MemoryStream objects, and forcing garbage collection doesn't seem to help.

Is this likely due to my usage of Mapsui, or had I better check the rest of my code? I'm not storing the layers anywhere else, so I'd think they would be disposed of. This is Mapsui 1.0.7 from Nuget. I'd appreciate any advice.

Thanks

**Comment by pauldendulk** [comment]
Thanks for you report. What platform are you using?

About the code sample: Layers.Clear() removes all layers from the map, so subsequent calls on Map.Dispose and Map.ClearCache have no effect. Btw. Map.Dispose should be removed from Mapsui. Just removing the layers should be enough to clear the memory. 

A complication is that there is a fetcher thread. It is in a waiting state when not fetching. Perhaps this is causing a leak.

Do you have the same problem with OsmLayer?

**Comment by pauldendulk** [comment]
The TileLayer.Abort has caused a thread leak in the past (which amounts to a memory leak) therefore the implementation was removed. This was in a scenario where one layer was reused for several TileSources. Perhaps it is necessary to abort when a layer is removed altogether. 

To investigate this you could create a copy of the TileLayer in your own code and make it so that TileFetcher.AbortFetch is called before removing this layer.

If this fixes it we know what the problem is and can work on a proper design for this.

**Comment by reubengann** [comment]
Thanks for your reply. This is on Windows 10 x64 and WPF. Yeah, I figured that clearing the layers should do the job, but when it didn't I kind of just took a kitchen sink approach.

Since I'm working with game maps, I haven't tried any OpenStreetMap implementations. I can't quickly adapt the code for that, but I can take a look at it and see if it behaves the same way.

Calling AbortFetch() on the TileLayer does not seem to make any difference. I'm now using
```
Mapsui.Layers.TileLayer tileLayer;

private void LoadMap(string location)
{
    if (tileLayer != null)
    {
        tileLayer.AbortFetch();
        MyMapControl.Map.Layers.Clear();                
    }                        
    _source = new MbTilesTileSource(new SQLiteConnectionString(PathLibrary[location], false));
    tileLayer = new Mapsui.Layers.TileLayer(_source);
    MyMapControl.Map.Layers.Add(tileLayer);
}
```

I observe no difference in the memory usage.

**Comment by pauldendulk** [comment]
In the Mapsui TileLayer AbortFetch is not implemented. You need to create your own copy of the Mapsui TileLayer and call _tileFetcher.AbortFetch() in TileLayer AbortFetch.

**Comment by reubengann** [comment]
That absolutely did the trick. I made a new class TileLayerAbort with
```
 public override void AbortFetch()
{
    _tileFetcher.AbortFetch();
}
```
Memory usage no longer increases monotonically, and stays quite manageable. Thanks. I'll leave it to you to close the issue.

**Comment by pauldendulk** [comment]
Good to hear the workaround works and thanks for testing this. I will fix this in the Mapsui before closing the issue.

**Comment by pauldendulk** [comment]
Fix published in 1.0.8
